### ISSUE TITLE
`svelte/no-navigation-without-resolve` false positive with Paraglide localizeHref (scaffolded by sv create)

### ISSUE THREAD

**Issue by playday3008** [sveltejs/eslint-plugin-svelte#1492]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.

### What version of ESLint are you using?

10.0.3

### What version of `eslint-plugin-svelte` are you using?

3.15.1

### What did you do?

```ts
// eslint.config.js
import prettier from 'eslint-config-prettier';
import path from 'node:path';
import { includeIgnoreFile } from '@eslint/compat';
import js from '@eslint/js';
import svelte from 'eslint-plugin-svelte';
import { defineConfig } from 'eslint/config';
import globals from 'globals';
import ts from 'typescript-eslint';
import svelteConfig from './svelte.config.js';

const gitignorePath = path.resolve(import.meta.dirname, '.gitignore');

export default defineConfig(
      includeIgnoreFile(gitignorePath),
      js.configs.recommended,
      ...ts.configs.recommended,
      ...svelte.configs.recommended,
      prettier,
      ...svelte.configs.prettier,
      {
              languageOptions: { globals: { ...globals.browser, ...globals.node } },
              rules: {
                      'no-undef': 'off'
              }
      },
      {
              files: ['**/*.svelte', '**/*.svelte.ts', '**/*.svelte.js'],
              languageOptions: {
                      parserOptions: {
                              projectService: true,
                              extraFileExtensions: ['.svelte'],
                              parser: ts.parser,
                              svelteConfig
                      }
              }
      }
);
```
```svelte
<!-- src/routes/+layout.svelte (generated by sv create) -->
<script lang="ts">
      import { page } from '$app/state';
      import { locales, localizeHref } from '$lib/paraglide/runtime';
      import './layout.css';
      import favicon from '$lib/assets/favicon.svg';

      let { children } = $props();
</script>

<svelte:head><link rel="icon" href={favicon} /></svelte:head>
{@render children()}

<div style="display:none">
      {#each locales as locale} // <-- also missing `(locale)` here, but that's manually very easy fixable
              <a href={localizeHref(page.url.pathname, { locale })}>{locale}</a>
      {/each}
</div>
```

### What did you expect to happen?

`localizeHref()` from `@inlang/paraglide-js` should be recognized as a valid navigation helper (like `resolve()`), or the scaffolded template should use a pattern that satisfies the rule. This is the default code generated by `sv create` with the Paraglide addon — it errors out of the box with no user modifications.

### What actually happened?

```
/src/routes/+layout.svelte
  15:6  error  Unexpected href link without resolve()  svelte/no-navigation-without-resolve
```

Wrapping with `resolve()` from `$app/paths` causes a TypeScript error because `localizeHref()` returns string, not the expected `` "/" | `/${string}/` `` type

### Link to **GitHub Repo** with Minimal Reproducible Example

```sh
bun x sv@0.12.5 create --template minimal --types ts --add prettier eslint vitest="usages:unit,component" playwright tailwindcss="plugins:typography,forms" sveltekit-adapter="adapter:auto" devtools-json paraglide="languageTags:en, es+demo:no" mcp="ide:other+setup:remote" --install bun .
```

**Comment by marekdedic** [comment]
Hi,
thanks for the issue. I'll need to look into this more, but first thoughts:
- `localizeHref` does not substitute kit's `resolve` - e.g. it doesn't prepend the base path. Therefore the rule seems to be correct in reporting this and you should use `resolve(localizeHref("..."))`
- `localizeHref` returns a string, not a literal type, but `resolve` expects a literal type - I'm not sure how to handle this, I'm only a casual user of paraglide. Could paraglide type the return value as a literal? If so, that would be the cleanest solution.

**Comment by sacrosanctic** [comment]
the addon generates the following code, we just need to modify it to return `ResolvedPathName` and the rest should fix itself. We shouldn't need to make any changes to the `+layout.svelte`.

```ts
/**
 * High-level URL localization function optimized for client-side UI usage.
 *
 * This is a convenience wrapper around `localizeUrl()` that provides features
 * needed in UI:
 *
 * - Accepts relative paths (e.g., "/about")
 * - Returns relative paths when possible
 * - Automatically detects current locale if not specified
 * - Handles string input/output instead of URL objects
 *
 * @see https://inlang.com/m/gerre34r/library-inlang-paraglideJs/i18n-routing
 *
 * @example
 * ```typescript
 * // In a React/Vue/Svelte component
 * const NavLink = ({ href }) => {
 *   // Automatically uses current locale, keeps path relative
 *   return <a href={localizeHref(href)}>...</a>;
 * };
 *
 * // Examples:
 * localizeHref("/about")
 * // => "/de/about" (if current locale is "de")
 * localizeHref("/store", { locale: "fr" })
 * // => "/fr/store" (explicit locale)
 *
 * // Cross-origin links remain absolute
 * localizeHref("https://other-site.com/about")
 * // => "https://other-site.com/de/about"
 * ```
 *
 * For server-side URL localization (e.g., in middleware), use `localizeUrl()`
 * which provides more precise control over URL handling.
 *
 * @param {string} href - The href to localize (can be relative or absolute)
 * @param {object} [options] - Options for localization
 * @param {Locale} [options.locale] - Target locale. If not provided, uses `getLocale()`
 * @returns {string} The localized href, relative if input was relative
 */
export function localizeHref(href, options) {
	const currentLocale = getLocale();
	const locale = options?.locale ?? currentLocale;
	const url = new URL(href, getUrlOrigin());
	const localized = localizeUrl(url, { locale });
	// if the origin is identical and the href is relative,
	// return the relative path
	if (href.startsWith('/') && url.origin === localized.origin) {
		// check for cross origin localization in which case an absolute URL must be returned.
		if (locale !== currentLocale) {
			const localizedCurrentLocale = localizeUrl(url, {
				locale: currentLocale
			});
			if (localizedCurrentLocale.origin !== localized.origin) {
				return localized.href;
			}
		}
		return localized.pathname + localized.search + localized.hash;
	}
	return localized.href;
}
```

**Comment by sacrosanctic** [comment]
maybe something like this
```diff
- * @param {string} href - The href to localize (can be relative or absolute)
+ * @param {import("$app/types").RouteIdWithSearchOrHash | import("$app/types").PathnameWithSearchOrHash} href - The href to localize (can be relative or absolute)
 * @param {object} [options] - Options for localization
 * @param {Locale} [options.locale] - Target locale. If not provided, uses `getLocale()`
- * @returns {string} The localized href, relative if input was relative
+ * @returns {import("$app/types").ResolvedPathname} The localized href, relative if input was relative */
export function localizeHref(href, options) {
	const currentLocale = getLocale();
	const locale = options?.locale ?? currentLocale;
	const url = new URL(href, getUrlOrigin());
	const localized = localizeUrl(url, { locale });
	// if the origin is identical and the href is relative,
	// return the relative path
	if (href.startsWith('/') && url.origin === localized.origin) {
		// check for cross origin localization in which case an absolute URL must be returned.
		if (locale !== currentLocale) {
			const localizedCurrentLocale = localizeUrl(url, {
				locale: currentLocale
			});
			if (localizedCurrentLocale.origin !== localized.origin) {
-				return localized.href;
+				return resolve(localized.href);
			}
		}
-		return localized.pathname + localized.search + localized.hash;
+		return resolve(localized.pathname + localized.search + localized.hash);
	}
-	return localized.href;
+	return resolve(localized.href);
}
```

**Comment by marekdedic** [comment]
Thanks, but I don't think localization should do resolving automatically - what if you want to localize an external href?

I think it's semi-reasonable that if you localize a valid route, the result is also a valid route (which is not actually true, you could make a typo in your localization and the route would then be invalid 🤷).

However, be aware that paraglide is not just a Svelte thing, but an independent i18n project. It might be possible to do something to integrate them, but you'd need to propose this to the SvelteKit & Paraglide teams (I'm not affiliated with either, FYI)

**Comment by sacrosanctic** [comment]
<img width="847" height="296" alt="Image" src="https://github.com/user-attachments/assets/5a98c40a-cdc5-4f61-a791-3a8c2f2d8c8c" />

@marekdedic Is this expected behaviour? I'm not getting linting error anymore

**Comment by marekdedic** [comment]
Hmm, that's weird. Can you look at the return type of `localizeHref`?

**Comment by sacrosanctic** [comment]
It returns strong.

**Comment by sacrosanctic** [comment]
Nvm, I cannot replicated it.

Regarding this issue, I think it can be closed. The fix we went with in the Svelte CLI is to just cast the output of `localizeHref`.

**Comment by marekdedic** [comment]
Thanks, but I think that's probably the wrong choice - you are hiding the genuine issue raised by this rule. If you know it's internal navigation, you should `resolve()` instead. If you know it's external navigation, you should `rel="external"`. If you don't know, then you should leave it to the user to decide which of these 2 is correct.

**Comment by sacrosanctic** [comment]
The alternative is the just reimplement the paraglide fn to return the correct type.

**Comment by marekdedic** [comment]
But that's the same problem - unless you know whether the navigation is internal or external, you don't know the correct return type any more specific than `string`...

**Comment by sacrosanctic** [comment]
@marekdedic with consideration about what you said, what about this implementation?

```ts
/**
 * A function to make both `paraglide` and `sveltekit` happy
 */
export const localize = <
	H extends (URL & { pathname: ResolvedPathname }) | PathnameWithSearchOrHash | `http${string}`
>(
	href: H,
	options?: { locale: Locale }
): H extends (URL & { pathname: ResolvedPathname }) | Pathname ? ResolvedPathname : string => {
	if (href instanceof URL || href.startsWith('/')) {
		const url = href instanceof URL ? href : new URL(href, 'http://example.com');
		const relative = url.pathname + url.search + url.hash;
		const localized = localizeHref(relative, options) as PathnameWithSearchOrHash;
		const resolved = resolve(localized);
		return resolved;
	}

	// @ts-expect-error `PathnameWithSearchOrHash` is type `string` and can't be typed narrowed
	return localizeHref(href, options);
};
```
It uses the svelte native url class type, otherwise, it checks the string for relative and absolute and return accordingly.

**Comment by marekdedic** [comment]
Thanks, I think the general idea is heading the correct way, but the logic is not exact. I think the `if` condition is incorrect. Look at the rule logic - it's actually pretty complicated and depends on where the value is - some places don't allow absolute URLs for example.
